### ISSUE TITLE
Webinterface rest and websocket

### ISSUE THREAD

**Issue by JackLeEmmerdeur** [zekroTJA/yuri2#9]
Hello again.

As you suggested in issue #8 to update go, my dummy-self couldn't find the time to update Go manually and managed it with this [helper](https://github.com/udhos/update-golang). The "undefined: sort.SliceStable" is gone and I possess a nice "yuri2" executable now.

What I've done:
- I've registered a bot in discord, obtained the client-id and secret.

In application.yaml:
- server.port: 2333
- server.address: 127.0.0.1
- Entered password in lavalink.server.password 

In config.yaml:
- api.client_id and api.client_secret from Bot-registration
- discord.owner_id in the format USERNAME#NUMBER
- lavalink.address: localhost:2333
- lavalink.password: the same as in application.yaml
- lavalink.sounds_locations configured

Started the Lavalink.jar: 
No error.

Started yuri: 
- No error.
- Found sounds, too
- Opened ws://localhost:2333
- Gave an invite link for the bot

I invited yuri2 to my discord. It entered the building.

Compiled the web-folder with ng.

When I started the development-server, the userinterface is shown on localhost:4200 but it fails to connect to rest and websockets-server on port 8080. Is that the wrong port maybe? I changed it meanwhile to 2333, for testing in rest.service.ts and ws.service.ts. Chrome is giving me this:

![Bildschirmfoto](https://user-images.githubusercontent.com/4656882/68998837-cd9edf80-08b7-11ea-9a3b-fdc8fcb1800c.png)

This all looks very promising execpt the last step and as soon as I'm able to broadcast the first sound in my server, I make the installation permanent with pm2 (I guess).

I'm also fine without a web-interface, if I could command yuri2 via discord.

Is the latter possible or can you imagine what the root cause of the failing rest and websocket servers is?

Thanks.

**Comment by yukinose** [comment]
try changing the ports in rest.service.ts and ws.service.ts to 4200 in your example

**Comment by JackLeEmmerdeur** [comment]
Ok now I do not get a 404 for "api/localsounds" anymore. The ws-endpoint localsounds seems to be found, but the webapp can't authenticate to the websocket (see at the bottom).

This is what I did:

- In **yuri2/bin/application.yaml**: server.port=4200
- In **yuri2/bin/config.yaml**: lavalink.address=localhost:4200
- In **yuri2/web/src/app/api/rest/rest.service.ts** in line 75:
       `this.rootURL = environment.production ? '' : 'http://localhost:4200';`
- In **yuri2/web/src/app/api/ws/ws.service.ts** in line 31:
       `: 'ws://localhost:4200/ws';`

Without changing port to 4200 in application.yaml and config.yaml, I get the 404 for "api/localsounds" again.

Results:
Lavalink starts on port 4200
yuri2 connects to Lavalink and loads the sound files.
ng serve on webapp runs the app on 8080

When opening localhost:8080 I get the UI and some ws and rest exception toast-errors.

Network looks like this:
![Bildschirmfoto-1](https://user-images.githubusercontent.com/4656882/70277556-e9105280-17b2-11ea-9bd2-888c86eb24dd.png)

The 4 errors roughly correspond to the toast-errors:

- 1st is a 401 Unauthorized for "http://localhost:4200/api/localsound"
  Response Header contains Lavalink-Api-Version: 3

- 2nd is a 401 for ws://localhost:4200/ws with message:
       HTTP Authentication failed; no valid credentials available

- 3rd is a 401 for http://localhost:4200/api/info

- 4th is the same as 2nd.

So how do I let the webapp know the credentials of the rest and websocket services?

Thanks for reading.

**Comment by yukinose** [comment]
i am not that familiar with the bot so i can't really give you much information about how it works (i'm not zekro). i have gotten the bot to work though and here are some steps i took:
- In **yuri2/bin/application.yaml**: server.port= 2333
- In **yuri2/bin/config.yaml**: lavalink.address=localhost:2333; api.address: 8080; api.public_address: http://localhost:8080
- In **yuri2/web/src/app/api/rest/rest.service.ts** in line 75: set port to 8080
- In **yuri2/web/src/app/api/ws/ws.service.ts** in line 31: set port to 8080

then, build the web server and move it into bin via the steps zekro provided in the previous issue:
```
$ cd web
$ npm install
$ npm run build --prod
$ cd ..
$ mkdir bin/web
$ cp -r web/dist/ bin/web/
```
after running the binary, you should be able to connect to the front end via localhost:8080. it should ask for "log in via discord", which is an oauth2 redirect... you will need to add http://localhost:8080/login/authorize  to your "OAuth2 Redirects" via: discord developer portal > [your application] > OAuth2. 

this should help you get it running. through this, the web interface is only accessible when the bot is running. it seems that the cookie generated by authenticating via oAuth2 is used for restApi authentication (see the "docs" folder for more information).

**Comment by JackLeEmmerdeur** [comment]
Thanks so much @yukinose and @zekroTJA for the detailed instructions. The webinterface is running now and is presenting the defined songs in config.yaml->lavalink.sound_locations.

I still stumbled on one showstopper:
After starting a song in the webinterface I get a "WS Request Error - command failed: unexpected end of JSON input". My bot plays nothing.

You can close this issue atm. I'll try to figure that out myself and annoy you again, if I won't figure it out.

**Comment by JackLeEmmerdeur** [comment]
Yay! Huge Success in my night long debugging session.

How I found the culprit codeline:

- Cloned Lavalink and started debugging with IntelliJ with my custom application.yaml
  - Found out lavalink.server > io >WebsocketHandlers.kt > fun play(...) isn't called which is
    called by lavalink.server > io > SocketServer.kt > handleTextMessageSafe(...) where websocket
    messages are received by yuri-go or the yuribot?
- So I debugged the go app with vscode to see where the "play" websocket message is issued
   - It's issued in the gavalink-lib in 
     vendor > github.com > zekroTJA > gavalink > player.go > func PlayAt(...) which in turn
     is initially called by the main go app in
     internal > player > player.go > func Play(...): line 360
   - When debugging the main Playfunc I found out that line 360 isn't reached
   - Instead the "unexpected JSON" error is catched in line 344 from p.loadTrack() which eventually
     is getting the error from gavalink > node.go in in
     func LoadTracks > line 184: `resp, err := http.DefaultClient.Do(req)`
     where the loadtracks Rest-endpoint "loadtracks" in AudioLoaderRestHandler.java 
     is called. The problem is the Rest-endpoint isn't called at all in lavalink
   - I figured out that http.DefaultClient.Do(req) requests
    "http://localhost:2333/loadtracks?identifier=/path/to/track"
    and /path/to/track contained spaces which weren't encoded.
    That was it!

TL;DR Fix:
In the go application in
- internal > player.go
  - Added `"net/url"` to imports
  - In line 344 exchanged
    `track, err := p.loadTrack(node, ident)`
    with
    `track, err := p.loadTrack(node, url.PathEscape(ident))`

It's working flawlessly now.

@zekroTJA The bot is the only discord soundbot working for me. The source code is really exceptionally clean and very well documented. A really huge effort to interconnect all the parts. I aspire to get to that level with rust or java, especially relating to oauth and websockets some day.
I think where this project could improve is an interactive setup application and if you think that is feasible you haven't anything done for that, I might volunteer for that.
